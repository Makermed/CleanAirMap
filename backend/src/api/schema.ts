import { makeSchema } from 'nexus'
import { join } from 'path'
import * as allTypes from './graphql'

export const schema = makeSchema({
  types: allTypes,
  outputs: {
    typegen: join(__dirname, '../..', 'nexus-typegen.ts'),
    schema: join(__dirname, '../..', 'schema.graphql'),
  },
  contextType: {
    module: join(__dirname, "./context.ts"),
    export: "Context",
  },
  sourceTypes: {
    headers: [ 'import { LocationModel } from "./src/dataAccess/dataTypes" '],
    modules: []
  }
})
