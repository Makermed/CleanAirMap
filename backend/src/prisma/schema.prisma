// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Location {
  locationId  Int      @id @default(autoincrement())
  name        String
  type        Int
  street      String
  locality    String?
  place       String
  district    String?
  region      String
  postcode    String?
  country     String
  latitude    Float
  longitude   Float
  description String?
  avgCo2      Float?
  rooms       Room[]
  created_id  Int
  created_at  DateTime @default(now())
}

model Room {
  roomId       Int       @id @default(autoincrement())
  locationId   Int
  location     Location  @relation(fields: [locationId], references: [locationId])
  name         String
  readings     Reading[]
  created_id   Int
  created_user User      @relation(fields: [created_id], references: [id])
  created_at   DateTime  @default(now())
}

model Reading {
  id           Int      @id @default(autoincrement())
  value        Decimal
  unit         String
  notes        String?
  locationId   Int
  readingRoom  Room?    @relation(fields: [roomId], references: [roomId])
  roomId       Int?
  created_id   Int
  created_user User     @relation(fields: [created_id], references: [id])
  created_at   DateTime @default(now())
}

model User {
  id       Int       @id @default(autoincrement())
  rooms    Room[]
  readings Reading[]
}
