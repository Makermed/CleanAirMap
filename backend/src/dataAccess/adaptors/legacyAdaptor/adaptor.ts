import { IDataAdaptor, ILocation, IReading, IRoom } from "../../interfaces";
import LocationDAO from "./LocationDAO";
import RoomDAO from "./RoomDAO";
import ReadingDAO from "./ReadingDAO";
import { HttpLink, ApolloClient, InMemoryCache, NormalizedCacheObject } from "@apollo/client/core";
import fetch from 'node-fetch'

/**
 * This adaptor queries from the legacy GraphQL endpoint and is just for
 * development/migration.
 */
export class LegacyAdaptor implements IDataAdaptor {
    client: ApolloClient<NormalizedCacheObject>;
    locationDAO: ILocation;
    roomDAO: IRoom;
    readingDAO: IReading;

    constructor(private endpoint: string) {
        this.client = new ApolloClient({
            link: new HttpLink({
                uri: endpoint,
                fetch: fetch
              }),
            cache: new InMemoryCache()
          });
      this.locationDAO = new LocationDAO(this.client);
      this.roomDAO = new RoomDAO(this.client);
      this.readingDAO = new ReadingDAO(this.client)
    }
  }
