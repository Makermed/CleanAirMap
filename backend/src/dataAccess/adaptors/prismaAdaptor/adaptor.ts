import { PrismaClient } from "@prisma/client";
import { IDataAdaptor, ILocation, IRoom } from "../../interfaces";
import { LocationDAO } from "./LocationDAO";
import { RoomDAO } from "./RoomDAO";

const prisma = new PrismaClient({
    log: ['query', 'info', 'warn', 'error'],
  })

export class PrismaDataAdaptor implements IDataAdaptor {
    client : PrismaClient
    locationDAO: ILocation
    roomDAO: IRoom

    constructor() {
        this.client = prisma
        this.locationDAO = new LocationDAO(this.client)
        this.roomDAO = new RoomDAO(this.client)
    }
}