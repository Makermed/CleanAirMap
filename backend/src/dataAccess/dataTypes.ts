export { User } from "@prisma/client"
export { Reading } from "@prisma/client"
export { Location } from "@prisma/client"
export { Room } from "@prisma/client"

// TODO: figure out if this can be referenced from Prisma types
export interface LocationModel {
    id: number
    name: String
    type: number
    street: String
    locality: String | null
    place: String
    district: String | null
    region: String
    postcode: String | null
    country: String
    latitude: number
    longitude: number
    description: String | null
    avgCo2: number | null
    created_id: number
    created_at: Date
}