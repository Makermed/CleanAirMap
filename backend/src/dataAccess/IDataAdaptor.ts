import { PrismaDataAdaptor } from "./adaptors/prismaAdaptor/adaptor";
import { ILocation, IReading, IRoom } from "./interfaces";

export interface IDataAdaptor {
    locationDAO : ILocation
    roomDAO : IRoom
    readingDAO : IReading
}