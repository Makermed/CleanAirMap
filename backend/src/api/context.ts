import { IDataAdaptor } from "../dataAccess/interfaces";
import { PrismaDataAdaptor } from "../dataAccess/adaptors/prismaAdaptor/adaptor";

export interface Context {
    db: IDataAdaptor
}

export const createContext =async () => ({
    db: new PrismaDataAdaptor()
})