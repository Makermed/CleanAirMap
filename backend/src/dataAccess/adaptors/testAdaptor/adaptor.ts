import { IDataAdaptor } from "../../IDataAdaptor";
import { ILocation } from "../../ILocation";
import { IRoom } from "../../IRoom";
import { LocationModel, RoomModel } from "../../dataTypes";

class TestLocationDAO implements ILocation {
    getById(id: number): Promise<LocationModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }

    getMany(maxResults?: number | undefined): Promise<LocationModel[]> {
        return Promise.reject(new Error("Not implemented"));
    }

    getRooms(locationId: number): Promise<RoomModel[] | null> {
        return Promise.reject(new Error("Not implemented"));
    }

    create(reading: LocationModel): Promise<LocationModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }
}

class TestRoomDAO implements IRoom {
    create(reading: RoomModel): Promise<RoomModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }
    getById(id: number): Promise<RoomModel | null> {
        return Promise.reject(new Error("Not implemented"));
    }
    getByLocation(location: LocationModel): Promise<RoomModel[] | null> {
        return Promise.reject(new Error("Not implemented"));
    }
    getMany(maxResults?: number | undefined): Promise<RoomModel[]> {
        return Promise.reject(new Error("Not implemented"));
    }
}

export class TestAdaptor implements IDataAdaptor {
    locationDAO: ILocation;
    roomDAO: IRoom;

    constructor() {
        this.locationDAO = new TestLocationDAO();
        this.roomDAO = new TestRoomDAO();
    }
}