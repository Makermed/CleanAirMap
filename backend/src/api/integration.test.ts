import { ApolloServer, GraphQLResponse } from "@apollo/server";
import { schema } from "./schema";
import { Context } from "./context"
import { TestAdaptor } from "../dataAccess/adaptors/testAdaptor/adaptor";
import { LocationModel, RoomModel } from "../dataAccess/dataTypes";
import { assert } from "console";

const testServer = new ApolloServer<Context>({schema})

const testContext = {
    db: new TestAdaptor()
}

function mockSingleLocation() : Promise<LocationModel | null> {
    const location : LocationModel = {
        locationId: 1,
        name: 'Location 1',
        type: 3,
        street: '1 Main St',
        locality: null,
        place: 'AnyTown',
        district: null,
        region: 'StateProv',
        postcode: 'H0H0H0',
        country: 'Any Nation',
        latitude: 30,
        longitude: 40.32908,
        description: 'A fun place to be',
        avgCo2: 900,
        created_id: 1,
        created_at: new Date("2023-01-01")

    }
    return Promise.resolve(location)
}

function mockSingleLocationRooms() : Promise<RoomModel[] | null> {
  const rooms : RoomModel[] = [{
    roomId : 1,
    locationId: 1,
    name: 'L1 Room 1',
    created_id: 1,
    created_at: new Date("2023-01-01")
  },{
    roomId : 2,
    locationId: 1,
    name: 'L1 Room 2',
    created_id: 1,
    created_at: new Date("2023-01-01")
  },{
    roomId : 3,
    locationId: 1,
    name: 'L1 Room 3',
    created_id: 1,
    created_at: new Date("2023-01-01")
  },]
  return Promise.resolve(rooms)
}

it('fetches a single location, no rooms', async() => {
    testContext.db.locationDAO.getById = jest.fn().mockReturnValueOnce(mockSingleLocation())

    const response = testServer.executeOperation(
        {
            query: SINGLE_LOCATION_NO_ROOMS_QUERY,
            variables: { locationId: 1}
        },
        {
            contextValue: testContext
        }
    )

    const result = await response

    assert(result.body.kind === 'single')
    expect(result.body).toMatchSnapshot()
  }
)

it('fetches a single location\'s rooms', async() => {
  testContext.db.locationDAO.getById = jest.fn().mockReturnValueOnce(mockSingleLocation())
  testContext.db.locationDAO.getRooms = jest.fn().mockReturnValueOnce(mockSingleLocationRooms())

  const response = testServer.executeOperation(
      {
          query: SINGLE_LOCATION_WITH_ROOMS_QUERY,
          variables: { locationId: 1}
      },
      {
          contextValue: testContext
      }
  )

  const result = await response

  assert(result.body.kind === 'single')
  expect(result.body).toMatchSnapshot()
}
)

const SINGLE_LOCATION_NO_ROOMS_QUERY =
`query Locations($locationId: Int) {
  locations(locationId: $locationId) {
    locationId
    name
    avgCo2
    created_at
    created_id
    description
    district
    locality
    place
    position
    postcode
    region
    street
    type
  }
}`

const SINGLE_LOCATION_WITH_ROOMS_QUERY =
`query Locations($locationId: Int) {
  locations(locationId: $locationId) {
    locationId
    name
    rooms {
      roomId
      locationId
      name
      created_id
      created_at
    }
  }
}`