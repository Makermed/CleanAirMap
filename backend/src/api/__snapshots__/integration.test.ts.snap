// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches a single location's rooms 1`] = `
{
  "kind": "single",
  "singleResult": {
    "data": {
      "locations": [
        {
          "locationId": 1,
          "name": "Location 1",
          "rooms": [
            {
              "created_at": 2023-01-01T00:00:00.000Z,
              "created_id": 1,
              "locationId": 1,
              "name": "L1 Room 1",
              "roomId": 1,
            },
            {
              "created_at": 2023-01-01T00:00:00.000Z,
              "created_id": 1,
              "locationId": 1,
              "name": "L1 Room 2",
              "roomId": 2,
            },
            {
              "created_at": 2023-01-01T00:00:00.000Z,
              "created_id": 1,
              "locationId": 1,
              "name": "L1 Room 3",
              "roomId": 3,
            },
          ],
        },
      ],
    },
    "errors": undefined,
  },
}
`;

exports[`fetches a single location, no rooms 1`] = `
{
  "kind": "single",
  "singleResult": {
    "data": {
      "locations": [
        {
          "avgCo2": 900,
          "created_at": 2023-01-01T00:00:00.000Z,
          "created_id": 1,
          "description": "A fun place to be",
          "district": null,
          "locality": null,
          "locationId": 1,
          "name": "Location 1",
          "place": "AnyTown",
          "position": {
            "coordinates": [
              40.32908,
              30,
            ],
          },
          "postcode": "H0H0H0",
          "region": "StateProv",
          "street": "1 Main St",
          "type": "CAFE",
        },
      ],
    },
    "errors": undefined,
  },
}
`;
